### DISCUSSION TITLE
Access bevy_ecs World in threads?

### DISCUSSION THREAD

**Discussion by supagu** [bevyengine/bevy#15486] [Q&A] [answered]
I am writing a chat bot and trying to use bevy_ecs with it. 
Every time the bot responds to a chat message, it is in a separate thread.

I want to access the World in these threads to be able to manipulate the world, how ever when I try to do this, I get an error:

```
future cannot be sent between threads safely
within `impl Future<Output = Result<(), ()>>`, the trait `Send` is not implemented for `std::sync::MutexGuard<'_, World>`, which is required by `impl Future<Output = Result<(), ()>>: Send`
```

I have the World wrapped in a Mutex:
```
pub static ref GLOBAL_WORLD: Mutex<World> = Mutex::new(World::default());
```
Full src: https://github.com/bit-shift-io/text-rpg/blob/main/src/globals.rs

Inside my thread function:
```
                let mut world = GLOBAL_WORLD.lock().unwrap(); // Error cause by this line.
                world.spawn(Room {});
 ```
 Full src: https://github.com/bit-shift-io/text-rpg/blob/main/src/commands/start_a_new_game.rs
 
 I haven't done much with rust multi threading so not sure what options I have to resolve this problem.
 Can anyone help?

**Comment by SkiFire13** [comment, marked as answer]
As the error message mentions, your problem is with the `MutexGuard` returned by `lock` not being `Send`, rather than `World`. You should be able to solve this by wrapping those two lines in a block, so that the `MutexGuard` is not in scope when the next `.await` is called.

**Comment by supagu** [comment]
That works, thanks!
